File: Day_10/Day10.cpp
#include <iostream>
#include <fstream>
#include <stack>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
using namespace std;

map<char, int> _errorScoreMap{
    {')', 3},
    {']', 57},
    {'}', 1197},
    {'>', 25137}};

map<char, int> _incompleteScoreMap{
    {'(', 1},
    {'[', 2},
    {'{', 3},
    {'<', 4}};

int getErrorScore(string s)
{
    stack<char> stk;

    for (char ch : s)
    {
        if (ch == '(' || ch == '[' || ch == '{' || ch == '<')
            stk.push(ch);
        else if (
            ((ch == ')' || ch == ']' || ch == '}' || ch == '>') && stk.empty()) //
            || (ch == ')' && !stk.empty() && stk.top() != '(')                  //
            || (ch == ']' && !stk.empty() && stk.top() != '[')                  //
            || (ch == '}' && !stk.empty() && stk.top() != '{')                  //
            || (ch == '>' && !stk.empty() && stk.top() != '<'))
            return _errorScoreMap[ch];
        else
            stk.pop();
    }
    return 0;
}

long getCompletionScore(string s)
{
    stack<char> stk;

    for (char ch : s)
    {
        if (
            ch == '('    //
            || ch == '[' //
            || ch == '{' //
            || ch == '<')
            stk.push(ch);
        else if (
            ch == ')'    //
            || ch == ']' //
            || ch == '}' //
            || ch == '>')
            stk.pop();
    }
    long score = 0;
    while (!stk.empty())
    {
        score = score * 5 + _incompleteScoreMap[stk.top()];
        stk.pop();
    }
    return score;
}

int main()
{
    ifstream inputFile("./Day10_Input.txt");
    string input;
    vector<string> lines;
    vector<string> incompleteLines;
    while (getline(inputFile, input))
    {
        if (!input.empty() && input[input.length() - 1] == '\n')
            input.erase(input.length() - 1);
        lines.push_back(input);
    }

    // Part 1
    int errorScore = 0;
    for (string line : lines)
    {
        int score = getErrorScore(line);
        if (score == 0)
            incompleteLines.push_back(line);
        errorScore += score;
    }
    cout << "Corruption Score: " << errorScore << endl;

    // Part 2
    vector<long> completionScores;
    int middleIndex;
    for (string line : incompleteLines)
    {
        completionScores.push_back(getCompletionScore(line));
    }
    sort(completionScores.begin(), completionScores.end());
    middleIndex = completionScores.size() / 2;
    cout << "Middle Completion Score: " << completionScores[middleIndex] << endl;

    return 0;
}

File: Day_10/Day10.py
_errorScoreMap = {')': 3, ']': 57, '}': 1197, '>': 25137}
_completionScoreMap = {'(': 1, '[': 2, '{': 3, '<': 4}


def getErrorScore(line):
    stack = []
    for ch in line:
        if ch in ['(', '[', '{', '<']:
            stack.append(ch)
        elif (len(stack) == 0 and ch in [')', ']', '}', '>'])\
                or (ch == ')' and len(stack) > 0 and stack[-1] != '(')\
                or (ch == ']' and len(stack) > 0 and stack[-1] != '[')\
                or (ch == '}' and len(stack) > 0 and stack[-1] != '{')\
                or (ch == '>' and len(stack) > 0 and stack[-1] != '<'):
            return _errorScoreMap[ch]
        else:
            stack.pop()
    return 0


def getCompletionScore(line):
    stack = []
    for ch in line:
        if ch in ['(', '[', '{', '<']:
            stack.append(ch)
        else:
            stack.pop()
    score = 0
    while len(stack) > 0:
        score = score * 5 + _completionScoreMap[stack[-1]]
        stack.pop()
    return score


def getInput():
    lines = []
    with open('./Day10_Input.txt') as inputFile:
        for line in inputFile:
            line = line.rstrip()
            lines.append(line)
    return lines


lines = getInput()
incompleteLines = []

# Part 1
errorScore = 0
for line in lines:
    score = getErrorScore(line)
    if score == 0:
        incompleteLines.append(line)
    errorScore += score
print("Corruption Score: " + str(errorScore))

# part 2
completionScores = []
for line in incompleteLines:
    completionScores.append(getCompletionScore(line))
completionScores.sort()
midIndex = (len(completionScores) - 1) / 2
print("Middle Completion Score: " + str(completionScores[int(midIndex)]))
